src/main/kotlin/app/revanced/patcher/Context.kt

```diff
@@ -0,0 +1,118 @@
+package app.revanced.patcher
+
+import app.revanced.patcher.apk.Apk
+import app.revanced.patcher.apk.ResourceFile
+import app.revanced.patcher.util.method.MethodWalker
+import org.jf.dexlib2.iface.Method
+import org.w3c.dom.Document
+import java.io.Closeable
+import java.io.InputStream
+import java.io.OutputStream
+import javax.xml.parsers.DocumentBuilderFactory
+import javax.xml.transform.TransformerFactory
+import javax.xml.transform.dom.DOMSource
+import javax.xml.transform.stream.StreamResult
+
+/**
+ * A common interface to constrain [Context] to [BytecodeContext] and [ResourceContext].
+ */
+sealed interface Context
+
+/**
+ * A context for the bytecode of an [Apk.Base] file.
+ *
+ * @param options The [PatcherOptions] of the [Patcher].
+ */
+class BytecodeContext internal constructor(options: PatcherOptions) : Context {
+    /**
+     * The list of classes.
+     */
+    val classes = options.apkBundle.base.bytecodeData.classes
+
+    /**
+     * Create a [MethodWalker] instance for the current [BytecodeContext].
+     *
+     * @param startMethod The method to start at.
+     * @return A [MethodWalker] instance.
+     */
+    fun toMethodWalker(startMethod: Method): MethodWalker {
+        return MethodWalker(this, startMethod)
+    }
+}
+
+/**
+ * A context for [Apk] file resources.
+ *
+ * @param options The [PatcherOptions] of the [Patcher].
+ */
+class ResourceContext internal constructor(options: PatcherOptions) : Context {
+    /**
+     * The current [ApkBundle].
+     */
+    val apkBundle = options.apkBundle
+
+    /**
+     * Open an [DomFileEditor] for a given DOM file.
+     *
+     * @param inputStream The input stream to read the DOM file from.
+     * @return A [DomFileEditor] instance.
+     */
+    fun xmlEditor(inputStream: InputStream) = DomFileEditor(inputStream)
+}
+
+/**
+ * Wrapper for a file that can be edited as a dom document.
+ *
+ * This constructor does not check for locks to the file when writing.
+ * Use the secondary constructor.
+ *
+ * @param inputStream the input stream to read the xml file from.
+ * @param outputStream the output stream to write the xml file to. If null, the file will be read only.
+ *
+ */
+class DomFileEditor internal constructor(
+    private val inputStream: InputStream,
+    private val outputStream: Lazy<OutputStream>? = null,
+    private val onClose: (() -> Unit)? = null,
+) : Closeable {
+    private var closed: Boolean = false
+
+    /**
+     * The document of the xml file.
+     */
+    val file: Document = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(inputStream)
```

Comment:
That is possible, but are you really sure that is a good idea?
Implementing document would require overriding 70 methods.

Reply:
> would require overriding 70 methods.

That shouldn't be the case. `Document` should be a class so overriding is not necessary.

Reply:
> That shouldn't be the case. `Document` should be a class so overriding is not necessary.

Nope. [Document](https://docs.oracle.com/javase/8/docs/api/org/w3c/dom/Document.html) is an interface.

Reply:
In that case the implementation of `Document` for XML files should be used.

Reply:
> In that case the implementation of `Document` for XML files should be used.

I'll quote @Axelen123:
> I haven't found a class you can just extend to accomplish this either, which is not surprising because parsing is done using static methods.

Reply:
The documentation of Document or simply the method that returns the Document can be inspected for implementations of the interface.

Reply:
> The documentation of Document or simply the method that returns the Document can be inspected for implementations of the interface.

It appears the `Document` implementation and its parser are both internal in the JDK.